### ISSUE TITLE
setState does not rerender 

### ISSUE THREAD

**Issue by skaraman** [developit/preact#829]
hey this might be a very special case scenario

I'm using preact to dev in cocos creator (runs on `editor-framework`)

When i call `setState` in a component it runs through changing the state but does not rerender. I walked through the code and found that it's because my component is being flagged `_dirty: true`, if I set this to false it rerenders perfectly.

**Comment by developit** [comment]
Correct - setState() triggers batched re-renders, which means one setState() call does not necessarily produce one render.  If you need 1:1 rendering in response to a state change, I would recommend calling `forceUpdate()`:

```js
this.setState({ foo: 'bar' });
this.forceUpdate();  // synchronously re-renders
```

**Comment by skaraman** [comment]
yeah i'm not sure because without hacking `_dirty` or using `forceUpdate` my component never rerenders, so i'm not sure what else to do to trigger it

using `forceUpdate` after every` setState` is more trouble than setting `this._dirty = false` once in `constructor`

**Comment by developit** [comment]
`this._dirty` gets set to `true` by `setState()`, setting it in the constructor will only work once.  Also that's an internal property, it's called `__d` or something when using the minified version of preact.

Can you give a bit more detail on how you're using Preact?  I'm wondering if this is a symptom of something else.

**Comment by skaraman** [comment]
just using webpack and babel to bundle, like i mentioned it's a panel in `editor-framework` based on this repo https://github.com/editor-framework/example-react
The strangest part is probably the hook to add it in (and it get's loaded into a shadow-dom)

```
const Preact = Editor.require('packages://preact-wrapper/node_modules/preact')

const MainPreviews = Editor.require('packages://preact-wrapper/panel/dist/bundle.js')


Editor.Panel.extend({
	template: `
		<div id="view">
		</div>
	`,

	$: {
		"view": "#view"
	},

	ready () {
		// Render the Preact component and inject it into the #view wrapper.
		Preact.render (
			Preact.createElement(MainPreviews),
			this.$view
		)
	}
})
```

regarding `setState` setting dirty, when the component is run through `renderComponent` it sets dirty back to false so that's why it's working for me to set it once

**Comment by developit** [comment]
If I'm following along correctly, I think what you're seeing is actually the intended behavior.  Components are always initialized in a dirty state because they can't have been rendered yet (dirty indicates a component is in need of render).  Any setState() calls made before first render will be batched into the already-queued render.

Random thought: are you setting `preact.options.debounceRendering` or `preact.options.syncComponents` at all?

**Comment by skaraman** [comment]
>Random thought: are you setting preact.options.debounceRendering or preact.options.syncComponents at all?

nope

**Comment by developit** [comment]
how are you checking if it rendered?

**Comment by skaraman** [comment]
visual confirmation - I have my preact app running in a panel from the editorframework which basically acts as its own webpage, it runs on electron so it has dev tools too, I can debug and put breakpoints in the jsx 

for example purposes - it's a button that runs a function from its class that `setState.({buttons: buttonArr})` and would expect to render additional buttons when clicked (this all works with changing _dirty or forceUpdate, doesn't work when leaving it up to to the batch render trigger)

**Comment by developit** [comment]
@skaraman Definitely something funky going on here for sure. Would you be up for a screenshare or something?  Happy to help get to the bottom of things. Ping me on [Slack](https://preact-slack.now.sh) if you have a few minutes.

**Comment by skaraman** [comment]
sorry but now i'm getting a different error that is blocking me from even working, I accidentally deleted my package.json and webpack.config.babel.js files, i re-wrote them and now the whole app fails to load - 

> index.js:49 Uncaught (in promise) DOMException: Failed to execute 'createElement' on 'Document': The tag name provided ('[object Object]') is not a valid name.(…)

**Comment by developit** [comment]
That's usually caused by an export error - somewhere you're doing `<Foo>` where `Foo` is an object instead of a function.

**Comment by skaraman** [comment]
Yeah thanks, we talked about it in the slack team. It was a strange issue 
```
export default class MyClass extends Component {}
```
Throws the error 
```
class MyClass extends Component {}

module.exports = MyClass
```
Works
